gridfs.go

```diff
@@ -0,0 +1,601 @@
+type GridFSOptions struct {
+	Name         string
+	Prefix       string
+	Database     string
+	MongoDBURI   string
+	DryRun       bool
+	DeleteOnSync bool
+	DeleteOnPush bool
+	DeleteOnPull bool
+	Verbose      bool
+}
+
+func (o *GridFSOptions) validate() error {
+	if (o.DeleteOnPush != o.DeleteOnPull) && o.DeleteOnSync {
+		return errors.New("ambiguous delete on sync options set")
+	}
+
+	return nil
+}
+
+type gridfsBucket struct {
+	opts   GridFSOptions
+	client *mongo.Client
+}
+
+func (b *gridfsBucket) normalizeKey(key string) string { return b.Join(b.opts.Prefix, key) }
+
+func (b *gridfsBucket) denormalizeKey(key string) string {
+	if b.opts.Prefix != "" && len(key) > len(b.opts.Prefix)+1 {
+		key = key[len(b.opts.Prefix)+1:]
+	}
+	return key
+}
+
+// NewGridFSBucketWithClient returns a new bucket backed by GridFS with the
+// existing Mongo client and given options.
+func NewGridFSBucketWithClient(ctx context.Context, client *mongo.Client, opts GridFSOptions) (Bucket, error) {
+	if client == nil {
+		return NewGridFSBucket(ctx, opts)
+	}
+
+	if err := opts.validate(); err != nil {
+		return nil, err
+	}
+
+	return &gridfsBucket{opts: opts, client: client}, nil
+}
+
+// NewGridFSBucket returns a bucket backed by GridFS with the given options.
+func NewGridFSBucket(ctx context.Context, opts GridFSOptions) (Bucket, error) {
+	if err := opts.validate(); err != nil {
+		return nil, err
+	}
+
+	client, err := mongo.NewClient(options.Client().ApplyURI(opts.MongoDBURI))
+	if err != nil {
+		return nil, errors.Wrap(err, "constructing client")
+	}
+
+	connctx, cancel := context.WithTimeout(ctx, time.Second)
+	defer cancel()
+
+	err = client.Connect(connctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "connecting")
+	}
+
+	return &gridfsBucket{opts: opts, client: client}, nil
+}
+
+func (b *gridfsBucket) Check(ctx context.Context) error {
+	if b.client == nil {
+		return errors.New("no client defined")
+	}
+
+	return errors.Wrap(b.client.Ping(ctx, nil), "contacting DB")
+}
+
+func (b *gridfsBucket) Exists(ctx context.Context, key string) (bool, error) {
+	grid, err := b.bucket(ctx)
+	if err != nil {
+		return false, errors.Wrap(err, "resolving bucket")
+	}
+
+	if err = grid.GetFilesCollection().FindOne(ctx, bson.M{"filename": b.normalizeKey(key)}).Err(); err == mongo.ErrNoDocuments {
+		return false, nil
+	} else if err != nil {
+		return false, errors.Wrap(err, "finding file")
+	}
+
+	return true, nil
+}
+
+func (b *gridfsBucket) Join(elems ...string) string { return consistentJoin(elems) }
+
+func (b *gridfsBucket) bucket(ctx context.Context) (*gridfs.Bucket, error) {
+	if err := ctx.Err(); err != nil {
+		return nil, errors.Wrap(err, "fetching bucket with canceled context")
+	}
+
+	gfs, err := gridfs.NewBucket(b.client.Database(b.opts.Database), options.GridFSBucket().SetName(b.opts.Name))
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	dl, ok := ctx.Deadline()
```

Comment:
Since this is a slightly unusual pattern (usually, we can just pass the context into the DB operations), can we add a little code comment that GridFS does not accept contexts, so setting the read/write deadlines here tries to emulate what a context would do if it errored?

Reply:
Done! I also updated the mongo driver (the newer version enables passing in contexts on non-streaming functions) and passed in contexts where I could.